core/src/main/scala/cats/Eval.scala

```diff
@@ -134,23 +136,57 @@ final case class Now[A](value: A) extends Eval[A] {
  * by the closure) will not be retained, and will be available for
  * garbage collection.
  */
-final class Later[A](f: () => A) extends Eval[A] {
-  private[this] var thunk: () => A = f
-
-  // The idea here is that `f` may have captured very large
-  // structures, but produce a very small result. In this case, once
-  // we've calculated a value, we would prefer to be able to free
-  // everything else.
-  //
-  // (For situations where `f` is small, but the output will be very
-  // expensive to store, consider using `Always`.)
-  lazy val value: A = {
-    val result = thunk()
-    thunk = null // scalastyle:off
-    result
-  }
+final class Later[A](f: () => A) extends Eval[A] with Externalizable {
+
+  // we use lock to prevent possible problems with someone else
+  // synchronizing on 'this'.
+  private[this] val lock = new Object()
+  @volatile private[this] var thunk: Function0[A] = f
+  private[this] var result: A = _
+
+  final def value: A =
+    if (thunk == null) {
+      result
```

Comment:
I think there might be a nasty corner case here where both `thunk` and `result` evaluate to `null` from a secondary threads perspective, but would definitely love to find out I'm wrong.
Consider:
- Thread A enters `value` method and inspects `thunk`, seeing `thunk` as non `null`
- Thread A enters synchronized block
- Thread A inspects `thunk` again, and sees it as non `null`
- Thread A enters else branch and evaluates `thunk`, assigning the calculated `A` to `result`
- Thread A assigns `null` to `thunk`
- Thread B now enters `value` method and inspects `thunk`
- Thread B sees `thunk` as `null` due to volatility. The value has been written in Thread A as `null`.
- Thread B returns the value of `result`

I _think_ that it is possible for the Thread B `result` variable to still point to a `null` value, briefly in this scenario because I don't think either the volatility of `thunk` or the `synchronized` block which this code block does not take place in, make any guarantees about which value Thread B might see as the value in `result`. If I'm wrong here it would be great to find out why as it would improve my understanding of the JVM.